### ISSUE TITLE
Option argument value

### ISSUE THREAD

**Issue by stevedlawrence** [scallop/scallop#53]
Is there a way to get an optional argument value? Something kindof like this:

``` scala
val foo = opt[Option[String]]()
```

The idea being that it would be nice to have three states:
1. The option isn't provided:
   
   ```
   ScallopConf(List())) { val foo = ... }
   foo.get should equal (None)
   ```
2. The option is provided, but without a value:
   
   ```
   ScallopConf(List("-f")) { val foo = ... }
   foo.get should equal (Some(None))
   ```
3. The option is provided with a value:
   
   ```
   ScallopConf(List("-f", "bar")) { val foo = ... }
   foo.get should equal (Some(Some(bar)))
   ```

I've played around with a few different things, but couldn't get anything to give me exactly that.

**Comment by Rogach** [comment]
Here's the way to do it with custom converter:

``` scala
test ("optional value - new converter way") {
  def optConv[A](conv: String => A)(implicit tt: TypeTag[Option[A]]) =
    new ValueConverter[Option[A]] {
      def parse(s: List[(String, List[String])]) = {
        s match {
          case Nil => Right(None)
          case (_, Nil) :: Nil => Right(Some(None))
          case (_, v :: Nil) :: Nil => Right(Some(Some(conv(v))))
          case _ => Left(Unit)
        }
      }
      val tag = tt
      val argType = ArgType.LIST
    }
  implicit def strOptConv = optConv(identity)
  def getcf(args: Seq[String]) = new ScallopConf(args) {
    val foo = opt[Option[String]]()
  }
  val conf1 = getcf(Nil)
  conf1.foo.get === None
  val conf2 = getcf(List("-f"))
  conf2.foo.get === Some(None)
  val conf3 = getcf(List("-f", "bar"))
  conf3.foo.get === Some(Some("bar"))
}
```

It's ugly, huge, but works. But we can do better :) ValueConverter currently has only `map` defined on it, but if we go further and define `flatMap` as follows:

``` scala
def flatMap[B](fn: A => Either[Unit, Option[B]])(implicit tt: TypeTag[B]) =
  new ValueConverter[B] { child =>
    def parse(s: List[(String,List[String])]) =
      parent.parse(s).right.flatMap {
        case None => Right(None)
        case Some(a) => fn(a)
      }
    val tag = tt
    val argType = parent.argType
  }
```

we'll be able to get it to be more concise:

``` scala
test ("optional value - flatMap way") {
  implicit def optConv[A](implicit tt: TypeTag[Option[A]], conv: ValueConverter[List[A]]) =
    conv.flatMap {
      case Nil => Right(None)
      case v :: Nil => Right(Option(Option(v)))
      case _ => Left(Unit)
    }
  def getcf(args: Seq[String]) = new ScallopConf(args) {
    val foo = opt[Option[String]]()
  }
  val conf1 = getcf(Nil)
  conf1.foo.get === None
  val conf2 = getcf(List("-f"))
  conf2.foo.get === Some(None)
  val conf3 = getcf(List("-f", "bar"))
  conf3.foo.get === Some(Some("bar"))
}
```

It seems to work, but I hit a bug (#54) while implementing it, so currently it doesn't work. I'll fix that now, and we'll get a better way of converting our converters :)

**Comment by stevedlawrence** [comment]
Very cool. This looks like it will work perfectly. Thanks!

**Comment by stevedlawrence** [comment]
Is there an easy way to fix the --help output when these optional args? With your solutions, I get

```
 -f, --foo <arg>...
```

Which isn't really correct. This converter only allows a single value, and it's optional, so it should look like

```
 -f, --foo [arg]
```

Is there a way to accomplish this?

**Comment by Rogach** [comment]
Hm. That's actually an extremely interesting question.

Try this modification to the converter (I haven't even compiled that, it may not work):

``` scala
val argType = ArgType.V("[" + _ + "]")
```

**Comment by Rogach** [comment]
Oh, I'm royally stupid. Of course it doesn't work.

**Comment by stevedlawrence** [comment]
It actually does work for the --help output. But parsing fails because of a MatchError in match/case because it's an unknown argtype.

**Comment by Rogach** [comment]
Well, I could give a workaround, but it will probably be 2 pages long. It would be easier just to add a minor adjustment to the library.

I think that combining the parsing information (the fact that value converter denotes a flag or a list) and output information (formatting) is bad. It may be better to split it into two separate fields, so you would be able to override it.

What do you think?

**Comment by stevedlawrence** [comment]
Sounds like the right solution to me.

**Comment by Rogach** [comment]
I implemented it, and pushed the snapshot (for scala 2.10.x):

```
resolvers += "sonatype snapshots" at "https://oss.sonatype.org/content/repositories/snapshots"
libraryDependencies += "org.rogach" %% "scallop" % "0.9.1~3~SNAPSHOT"
```

**Comment by stevedlawrence** [comment]
Perfect! Works for me. Thanks again.
